client/cli/src/lib.rs

```diff
@@ -161,637 +127,200 @@ fn base_path(
-				Ok(n) => BlockId::number((n as u32).into()),
-				Err(_) => return Err(error::Error::Input("Invalid hash or number specified".into())),
-			}
-		};
-
-		let start = std::time::Instant::now();
-		let check = builder(config)?
-			.check_block(block_id);
-		let mut runtime = tokio::runtime::Runtime::new().unwrap();
-		runtime.block_on(check)?;
-		println!("Completed in {} ms.", start.elapsed().as_millis());
-		Ok(())
-	}
-}
+/// A helper function that:
+/// 1.  initialize
+/// 2.  runs any of the command variant of `CoreParams`
+pub fn run<F, G, E, FNL, FNF, B, SL, SF, BC, BB>(
+	mut config: Configuration<G, E>,
+	core_params: CoreParams,
+	new_light: FNL,
+	new_full: FNF,
+	spec_factory: F,
+	builder: B,
+	version: &VersionInfo,
+) -> error::Result<()>
+where
+	F: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
+	FNL: FnOnce(Configuration<G, E>) -> Result<SL, sc_service::error::Error>,
+	FNF: FnOnce(Configuration<G, E>) -> Result<SF, sc_service::error::Error>,
+	B: FnOnce(Configuration<G, E>) -> Result<BC, sc_service::error::Error>,
+	G: RuntimeGenesis,
+	E: ChainSpecExtension,
+	SL: AbstractService + Unpin,
+	SF: AbstractService + Unpin,
+	BC: ServiceBuilderCommand<Block = BB> + Unpin,
+	BB: sp_runtime::traits::Block + Debug,
+	<<<BB as BlockT>::Header as HeaderT>::Number as std::str::FromStr>::Err: std::fmt::Debug,
+	<BB as BlockT>::Hash: std::str::FromStr,
+{
+	init(&mut config, spec_factory, core_params.get_shared_params(), version)?;
 
-/// Command ready to purge the chain.
-pub struct ParseAndPreparePurge<'a> {
-	params: PurgeChainCmd,
-	version: &'a VersionInfo,
+	core_params.run(config, new_light, new_full, builder, version)
 }
 
-impl<'a> ParseAndPreparePurge<'a> {
-	/// Runs the command and purges the chain.
-	pub fn run<G, E, S>(
-		self,
-		spec_factory: S
-	) -> error::Result<()> where
-		S: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
-		G: RuntimeGenesis,
-		E: ChainSpecExtension,
-	{
-		let mut config = create_config_with_db_path::<(), _, _, _>(
-			spec_factory,
-			&self.params.shared_params,
-			self.version,
-			None,
-		)?;
-		fill_config_keystore_in_memory(&mut config)?;
-		let db_path = match config.database {
-			DatabaseConfig::Path { path, .. } => path,
-			_ => {
-				eprintln!("Cannot purge custom database implementation");
-				return Ok(());
-			}
-		};
-
-		if !self.params.yes {
-			print!("Are you sure to remove {:?}? [y/N]: ", &db_path);
-			stdout().flush().expect("failed to flush stdout");
-
-			let mut input = String::new();
-			stdin().read_line(&mut input)?;
-			let input = input.trim();
+/// Initialize substrate and its configuration
+///
+/// This method:
+///
+/// 1.  set the panic handler
+/// 2.  raise the FD limit
+/// 3.  initialize the logger
+/// 4.  update the configuration provided with the chain specification, config directory,
+///     information (version, commit), database's path, boot nodes and telemetry endpoints
+pub fn init<G, E, F>(
```

Comment:
I'm not a fan of this API as it's way too error-prone, but I guess we can improve it later.
As an example of how it's error-prone, I have absolutely no idea whether you forgot to set a field of the configuration in this function except by carefully reviewing. If a new field is added, it is totally possible that one forgets to update this function.

Reply:
Most defaults actually come from `Configuration::default()` but the problem is that some parameters need to be present to run a node (like `chain_spec` but others too) and that should fail with an assert somehow.

If a new field is added, one will have to update `Configuration::default()` with a default value. But if that value can only be provided after parsing, then it falls into the same problem of `chain_spec` and the other parameters that are Option while they really are important for running.

I don't see how to improve this at the moment.